Edit developer guide (#140)
CS2103-AY1819S1-W14-1/main: docs/DeveloperGuide.adoc

// end::dataencryption[]

// tag::login[]
=== Login

> ClinicIO ensure that proper user access rights is given and authenticated by login to ClinicIO.

Reviewer: ```suggestion
ClinicIO ensures appropriate user access rights are issued by authenticating users on login.
```

==== Current Implementation

> The login mechanism uses `UniqueDoctorList` to retrieve a list of doctors or receptionist based on user role. +

Reviewer: ```suggestion
The login mechanism uses `UniqueDoctorList` to retrieve a list of doctors or receptionist based on the user's role. +
```

==== Design Considerations

> ===== Aspect: How login executes

Reviewer: ```suggestion
===== Aspect: Login execution
```

// end::dataencryption[]

// tag::login[]
=== Login

ClinicIO ensures appropriate user access rights are issued by authenticating users at login.

==== Current Implementation

The login mechanism facilitated by using `UniqueDoctorList` or `UniqueReceptionistList` to retrieve a list of doctors or receptionist based on the user's role. +

These operations are exposed in the Model interface as Model#hasDoctor() and Model#getDoctor() for doctors
or Model#hasReceptionist() and Model#getReceptionist() for receptionists respectively.

Given below is an scenario to illustrate the login authentication at each step: +

> . The user executes the `login` command by specifying the role, name and password. The `login` command determine the type of role that the user entered.

Reviewer: ```suggestion
. The user executes the `login` command by specifying the role, name and password. The `login` command determines the type of role entered by the user.
```

. If the role is doctor,

.. The `login` command call `Model#hasDoctor()` to ensure that this doctor exist inside ClinicIO.
.. Once ClinicIO check if user exists, the `login` command call `Model#getDoctor` to retrieve the doctor found in `UniqueDoctorList`.

> .. The `login` command verify password using `HashUtil#verifyPassword()` with the given doctor's password and password entered by user.

Reviewer: ```suggestion
.. The `login` command verifies a password using `HashUtil#verifyPassword()` with the given doctor's password and password entered by the user.
```

> .. Once verification has been made successfully, the `login` command raise an `SuccessfulLoginEvent` to indicate that a successful attempt to login has been made.

Reviewer: ```suggestion
.. To indicate successful completion of verification, the `login` command raises a `SuccessfulLoginEvent`.
```

. If the role is receptionist,

> .. the `login` command call `Model#hasReceptionist()` to ensure that this receptionist exist inside ClinicIO.

Reviewer: ```suggestion
.. the `login` command invokes `Model#hasReceptionist()` to ensure that this receptionist exists inside ClinicIO.
```

> .. Once ClinicIO check if user exists, the `login` command call `Model#getReceptionist()` to retrieve the receptionist found in `UniqueReceptionistList`.

Reviewer: ```suggestion
.. Once ClinicIO has checked that the user exists, the `login` command invokes `Model#getReceptionist()` to retrieve the receptionist found in `UniqueReceptionistList`.
```

> .. The `login` command verify password using `HashUtil#verifyPassword()` with the given receptionist's password and password entered by user.

Reviewer: ```suggestion
.. The `login` command verifies a password using `HashUtil#verifyPassword()` with the given receptionist's password and password entered by the user.
```

. Once verification has been made successfully, the `login` command raise an `SuccessfulLoginEvent` to indicate that a successful attempt to login has been made.

The following sequence diagram shows how the login authentication works:

==== Design Considerations

===== Aspect: Retrieval of password

> An alternative is to open another alert window to prompt user for password.

Reviewer: ```suggestion
An alternative is to open an additional alert window to prompt the user for a password.
```

> * Pros: More secure as the entire password will be censored.

Reviewer: ```suggestion
* Pros: Greater security as the entire password will be censored.
```

* Cons: Require a new window to be opened.
This will consume time depending on the system.

> In addition, This will take away the user experience.

Reviewer: ```suggestion
In addition, this will detract from the user experience due to the shift in focus required.
```
